### PR TITLE
Added financial aid calculator

### LINKED ISSUE #1097: Financial Aid: Cost Calculator
For programs with financial aid, learners will have to fill out a form to calculate their personal price. 

image

##  Acceptance criteria

- [ ] check with @roberthouse54 about whether country of residence appear in form
- [ ] This form will interact with the APIs implemented in #1053 
- [ ] user can bail out of the form using the "skip this and pay full price" link
- [ ] user cannot submit the form without checking the "I testify..." box

### PR DESCRIPTION
**Pull request by aliceriot**
#### What are the relevant tickets?

closes #1097 

#### What's this PR do?

This adds the financial aid calculator. This comprises:

- function to hit the REST API
- new reducer for financial aid stuff (and actions)
- UI components to render the calculator
- validation code for the calculator form
- a little bit of code to deal with countries -> currencies

#### Where should the reviewer start?

Read over the changes?

#### How should this be manually tested?

You should be able to go `/dashboard` and click 'Calculate your cost' to open the calculator. Fill in a salary amount and select `USD` ('United States Dollars') from the currency list. Then submitting the form should create a financial aid object for your user and program combo.

#### Screenshots (if appropriate)

calculator

### CODE REVIEW

**Comment by aliceriot** [comment]
@noisecapella

**Comment by noisecapella** [comment]
Can you look into this?

screenshot from 2016-09-28 11-35-05

**Comment by noisecapella** [comment]
We should filter out non-numeric data from the income field, just in case they put `$123456` and they get an error, for example

File: static/js/containers/FinancialAidCalculator.js [hunk 1]
```diff
@@ -0,0 +1,208 @@
+import React from 'react';
```
Comment (noisecapella) [inline static/js/containers/FinancialAidCalculator.js]: Can we add `// @flow` here or is flow being difficult?

File: static/js/reducers/financial_aid.js [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+  programId?:   number,
+  validation?:  FinancialAidValidation,
+};
+
+export const financialAid = (state: FinancialAidState = INITIAL_FINANCIAL_AID_STATE, action: Action) => {
+  switch (action.type) {
+  case START_CALCULATOR_EDIT:
+    return { ...FINANCIAL_AID_EDIT, programId: action.payload };
+  case CLEAR_CALCULATOR_EDIT:
+    return {};
```
Comment (noisecapella) [inline static/js/reducers/financial_aid.js]: Can you change to `return INITIAL_FINANCIAL_AID_STATE` to be explicit?

File: static/js/containers/FinancialAidCalculator.js (hunk 1 above)
Comment (aliceriot) [inline static/js/containers/FinancialAidCalculator.js]: haha it was being super difficult

File: static/js/reducers/financial_aid_test.js [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+      });
+      assert.deepEqual(state, expectation);
+    });
+  });
+
+
+  it('should let you clear edits', () => {
+    store.dispatch(startCalculatorEdit(1));
+    return dispatchThen(clearCalculatorEdit(), [CLEAR_CALCULATOR_EDIT]).then(state => {
+      assert.deepEqual(state, {});
```
Comment (noisecapella) [inline static/js/reducers/financial_aid_test.js]: Can you change to `assert.deepEqual(state, INITIAL_FINANCIAL_AID_STATE)` to be explicit?

File: static/js/containers/FinancialAidCalculator.js (hunk 1 above)
Comment (aliceriot) [inline static/js/containers/FinancialAidCalculator.js]: the problem is the `SelectField` which has type annotations which explicitly expect profile objects.

Comment (aliceriot) [inline static/js/containers/FinancialAidCalculator.js]: (and functions like `updateProfile: (p: Profile) => void`)

File: static/js/reducers/financial_aid_test.js [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    });
+  });
+
+  it('should let you update an edit in progress', () => {
+    store.dispatch(startCalculatorEdit(1));
+    let update = store.getState().financialAid;
```
Comment (noisecapella) [inline static/js/reducers/financial_aid_test.js]: We should make a copy here so we don't modify the original state

**Comment by aliceriot** [comment]
I agree re: non-numeric data. I implemented that for the date fields, i'll try to get the same thing going.

File: static/js/util/currency.js [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+import cc from 'currency-codes';
+import R from 'ramda';
+
+const codeToOption = code => (
+  { value: code, label: cc.code(code).currency }
+);
+
+const labelSort = R.sortBy(R.compose(R.toLower, R.prop('label')));
+
+const excludeUnusedCodes = R.reject(R.flip(R.contains)(['USS', 'USN']));
```
Comment (noisecapella) [inline static/js/util/currency.js]: Just curious, what codes are these?

Comment (aliceriot) [inline static/js/util/currency.js]: oh yeah, they are alternative codes for `USD`. `USN` is united states dollar next-day funds, and I think `USS` is some variant of that.

File: static/js/util/validation.js [hunk 6]
```diff
@@ -30,7 +34,8 @@ let isNilOrEmptyString = (val: any): boolean => (
   val === null || val === undefined || val === ""
 );
 
-let checkFieldPresence = (profile, requiredFields, messages: any): ValidationErrors => {
+// let checkFieldPresence = (profile, requiredFields, messages: any): ValidationErrors => {
```
Comment (noisecapella) [inline static/js/util/validation.js]: Remove this line please

File: static/js/util/currency.js (hunk 5 above)
Comment (aliceriot) [inline static/js/util/currency.js]: https://en.wikipedia.org/wiki/ISO_4217#USD.2FUSS.2FUSN.2C_three_currency_codes_belonging_to_the_US

File: static/js/util/validation.js [hunk 7]
```diff
@@ -299,3 +304,16 @@ export function combineValidators(...validators: Array<Function>): Function {
...
   ));
 }
+
+export function validateFinancialAid(edit: FinancialAidState): FinancialAidValidation {
+  let messages = {
+    'income': 'Income is required',
+    'currency': 'Please select a currency',
+  };
+  let required = Object.keys(messages).map(k => [k]);
+  let errors: FinancialAidValidation = checkFieldPresence(edit, required, messages);
```
Comment (noisecapella) [inline static/js/util/validation.js]: Hooray code reuse

File: static/js/containers/FinancialAidCalculator.js (hunk 1 above)
Comment (noisecapella) [inline static/js/containers/FinancialAidCalculator.js]: I see... we should fix that at some point, but fine to leave it for now

File: static/js/containers/FinancialAidCalculator.js [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+  validateFinancialAid,
+  sanitizeNumberString
+} from '../util/validation';
+import type { ProgramEnrollment } from '../flow/enrollmentTypes';
+import type {
+  FinancialAidState,
+  FinancialAidValidation,
+} from '../reducers/financial_aid';
+
+const currencySelect = (update, current) => (
```
Comment (noisecapella) [inline static/js/containers/FinancialAidCalculator.js]: We are not specifying a default currency based on the user's country of residence, right? Just making sure

Comment (aliceriot) [inline static/js/containers/FinancialAidCalculator.js]: yea that isn't implemented yet, USD is also the only thing supported right now

**Comment by noisecapella** [comment]
Functionality looks great

File: static/js/containers/FinancialAidCalculator.js (hunk 1 above)
Comment (aliceriot) [inline static/js/containers/FinancialAidCalculator.js]: I agree, but I didn't want to tackle it here

File: static/js/util/validation.js (hunk 7 above)
Comment (aliceriot) [inline static/js/util/validation.js]: :D

Comment (aliceriot) [inline static/js/util/validation.js]: All I had to do was get rid of `ValidationErrors` as the return type!

**Comment by aliceriot** [comment]
@noisecapella I think I addressed everything!

**Comment by noisecapella** [comment]
:+1:

**Pull request merged**